Add (EU-based) Honeywell evohome CH/DHW controller

zxdavb opened the pull request:
## Honeywell evohome CH/DHW Controllers (EU/EMEA only)

This is a hub component (for it's slave heating zones, with/without a DHW controller, a.k.a. boiler), and a climate entity.  Thus implemented as a evohome component.

No support for heating zones, DHW controller as yet - this will be added in future PRs.

## Description:
Support for a EU/EMEA-based Honeywell evohome controllers accessible via [mytotalconnectcomfort.com](mytotalconnectcomfort.com)  Such systems consists of a 'location' with a temperature control system (TCS, controller) with 0-12 heating zones (e.g. TRVs, UFH relays) and, optionally, a DHW controller.

To be clear, this component uses the **evohomeclient** client library, and _does not_ (can not) support US-based systems (`honeywell.py` uses two distinct client libraries, **evohomeclient**, and **somecomfort**).

**Related issue (if applicable):**
Addresses many of the limitations of the existing `honeywell.py` component:
- exposes the controller, and provides access to its operating modes (e.g. proper Away mode support)
- permits access to locations other than location 0 (or, more simply, you have multiple controllers - you can choose only one, though)
- TBA: exposes the DHW controller, and exposes its operating modes, etc.
- TBA: allows access to higher-precision `current_temperatures` via the v1 client library (0.1 displayed, 0.01 logged)
- significant improvements to efficiency (e.g. batching `add_entities()`, and initial update of state delayed until after `EVENT_HOMEASSISTANT_START`)
- many fewer api calls / `scan_interval` - it seems you can safely use a `scan_interval` of 60 seconds
- TBA: access to schedules & setpoints (read-only for now)

This component uses v2 of the **evohomeclient** client library (`honeywell.py` uses v1 only), with many new features/bug fixes:
- uses newer api, exposing more information
- resolve LOGGER bug (via a workaround in the component)
- much better exception handling

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#6252

## Example entry for `configuration.yaml` (if applicable):
```yaml
evohome:
  username: !secret evohome_username
  password: !secret evohome_password

# scan_interval: 300     # seconds, you can probably get away with 60
# location_idx: 0        # if you have more than 1 location, use this
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [ N/A ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

zxdavb committed:
Add support for Honeywell evohome CH/DHW systems

More flake8 corrections

Passes Flake8 tests

Almost passed flake8.pylint!

Passed all tox tests

Now it needs serious testing!

Almost ready to submit

BUGFIX: DHW state now functional

More improvements to available()

Solved the DHW temp units problem!

Last minute bug squash

to improve dicts merge

Trying to rebase

fixing more rbase errors

revert to creating HTTP_error_code internally for now

ready to submit PR

Added support for Honeywell evohome CH/DHW systems

homeassistant commented:
Hi @zxdavb, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#16427).

Once you do that we will be able to review and accept this pull request.

Thanks!

zxdavb commented:
There are three architectural issues with this PR:
1. The Climate entity is under discussion
2. The controller implemented as a Climate-like entity (there seems to be no appropriate component class).
3. The DHW controller is implemented as a Climate-like entity (there is no Boiler component class).

zxdavb committed:
Updated requirements_test_all.txt

zxdavb committed:
Fix: D401 First line should be in imperative mood

zxdavb committed:
Remove _LOGGER.info (replace with _LOGGER.debug)

zxdavb committed:
raise PlatformNotReady when RequestException during setup()

zxdavb committed:
Revert some LOGGER.debug to LOGGER.warning

zxdavb committed:
Improved logging levels, and removed some unwanted comments

zxdavb committed:
Merge branch 'dev' of https://github.com/home-assistant/home-assistant into evohome-component

zxdavb committed:
Improvments to logging - additional self._status info

zxdavb committed:
Merge branch 'dev' of https://github.com/home-assistant/home-assistant into evohome-component

zxdavb committed:
BUGFIX: DHW wrongly showing available = False (and some other tweaks)

zxdavb committed:
Fix trailing whitespace

zxdavb committed:
Merge branch 'dev' of https://github.com/home-assistant/home-assistant into evohome-component

frenck commented:
Could not find a matching documentation PR. Added the `docs-missing` label.

balloob commented on homeassistant/components/evohome.py:
```diff
@@ -0,0 +1,1941 @@
+
+API_VER = '0.2.7'  # alternatively, '0.2.5' is the version used elsewhere in HA
+
+if API_VER == '0.2.7':  # these vars for >=0.2.6 (is it v3 of the api?)...
```
This is very weird and should be removed. It should just be 0.2.7.

zxdavb replied:
This will be taken out.

balloob commented on homeassistant/components/evohome.py:
```diff
@@ -0,0 +1,1941 @@
+        return curr_op
+
+    @property
+    def state_attributes(self):
```
You are not allowed to override `state_attributes`. You can override `device_state_attributes`

zxdavb replied:
I'd had this bit in for some deep-level debugging - not needed now, so has been taken out with latest commit.

balloob commented:
This PR is too big. I can't get myself to review 2000 lines PRs. You should start with the least amount of code possible just to get the foundation in place. Maybe exclude zones and boiler support.

zxdavb committed:
Merge branch 'dev' of https://github.com/home-assistant/home-assistant into evohome-component

zxdavb committed:
Remove state_attributes override and API_VER code

zxdavb committed:
Removed heating zones, DHW and heuristics to reduce # lines of code

zxdavb commented:
I have reduced from ~2,000 lines to about 700!  I removed Heating zones, and the DHW controller.  I removed the heuristics code.  

I have changed the Controller (the only remaining component) to a `ClimateDevice`, and it now has a `current_temperature`, and a `target_temperature`.  

Some of the code may look a little 'strange' and I hope you'll forgive this.  The main reason for this is because I intend to restore some functionality over time.  For example:
```python
    @property
    def precision(self):
        """Return the temperature precision to use in the frontend UI."""
        return PRECISION_TENTHS
```
This is because the different 'slave' entities can have their own, distinct, `precision`.  Also:
```python
        if self._type & EVO_MASTER:
            curr_op = self._status['systemModeStatus']['mode']
        elif self._type & EVO_ZONE:
            curr_op = self._status[SETPOINT_STATE]['setpointMode']
        elif self._type & EVO_DHW:
            curr_op = self._status['stateStatus']['mode']
```

zxdavb committed:
Removed some more lines of code

zxdavb committed:
Removed unused configuration parameters

zxdavb commented:
Documentation PR added at home-assistant/home-assistant.io#6252

zxdavb committed:
Remove some more un-needed lines

zxdavb committed:
Merge branch 'dev' of https://github.com/home-assistant/home-assistant into evohome-component

zxdavb committed:
Removed more (uneeded) lines of code & fixed two minor typos

zxdavb committed:
Improvements to debug logging of available() = False

zxdavb committed:
Merge branch 'dev' of https://github.com/home-assistant/home-assistant into evohome-component

zxdavb committed:
Merge branch 'dev' of https://github.com/home-assistant/home-assistant into evohome-component

zxdavb committed:
Merge branch 'dev' of https://github.com/home-assistant/home-assistant into evohome-component

jeradM commented on homeassistant/components/evohome.py:
```diff
@@ -0,0 +1,690 @@
+            loc['locationInfo']['locationOwner'] = 'REDACTED'
+            loc[GWS][0]['gatewayInfo'] = 'REDACTED'
+
+# 3a.
```
These comments are getting out of hand. What does `3a` mean to anyone but you? Also, having all of these 0-indent comment lines makes your code really hard to read

zxdavb replied:
Removed.

zxdavb replied:
Removed all 0-indent comments, except those at the start of the code.

jeradM commented on homeassistant/components/evohome.py:
```diff
@@ -0,0 +1,690 @@
+# create a listener for the above...
+    hass.bus.listen(EVENT_HOMEASSISTANT_START, _first_update)
+
+# ... then finally, load the platform...
```
Again, too many comments. Do we really need a comment stating "load the platform" before calling the `load_platorm` function?

zxdavb replied:
removed

jeradM commented on homeassistant/components/evohome.py:
```diff
@@ -0,0 +1,690 @@
+    @callback
+    def _connect(self, packet):
+        """Process a dispatcher connect."""
+#       _LOGGER.debug("_connect(%s): got packet %s", self._id, packet)
```
Remove these commented log statements

zxdavb replied:
Done

jeradM commented on homeassistant/components/evohome.py:
```diff
@@ -0,0 +1,690 @@
+    @property
+    def target_temperature(self):
+        """Return the average target temperature of the Heating/DHW zones."""
+        num_temps = total_temp = 0
```
```python
temps = [zone['setpointStatus']['targetHeatTemperature'] 
         for zone in self._status['zones']]

avg_temp = sum(temps) / len(temps) if len(temps) > 0 else None
```

zxdavb replied:
I never used Python before this project - every day I lean something new!

Done

jeradM commented on homeassistant/components/evohome.py:
```diff
@@ -0,0 +1,690 @@
+# these are specific to this component
+CONF_LOCATION_IDX = 'location_idx'
+
+REQUIREMENTS = ['evohomeclient==0.2.7']
```
Move to directly below imports

zxdavb replied:
Done

zxdavb commented on homeassistant/components/evohome.py:
```diff
@@ -0,0 +1,690 @@
+    @property
+    def current_temperature(self):
+        """Return the average current temperature of the Heating/DHW zones."""
+        num_temps = total_temp = 0
```
Done

zxdavb committed:
Merge branch 'dev' of https://github.com/home-assistant/home-assistant into evohome-component

zxdavb committed:
Improvements to code, and code clean-up

zxdavb commented:
@jeradM Thanks for your input - I dunno if you're up for another look?

zxdavb committed:
Corrected a minor typo

zxdavb committed:
A small tidy up

zxdavb committed:
reduces precision of emulated temps floats to 0.1

jeradM commented on homeassistant/components/evohome.py:
```diff
@@ -0,0 +1,667 @@
+
+
+# these are for the controller's opmode/state and the zone's state
+EVO_RESET = 'AutoWithReset'
```
Please see https://github.com/home-assistant/architecture/issues/22. Are all of these possible states? We can't just have different states for a single climate platform

zxdavb replied:
I have previously read (and contributed to) that conversation.  Yes, they are all possible states, as reported by the client api.

The minutae is a little less simple than that, and I can do through the detail if you wish.

I have removed some CONSTANTS that applies only to the zones.

zxdavb replied:
@jeradM AM I OK to leave it as is?

BTW, this may give you an idea of the nature of it: http://www.automatedhome.co.uk/wp-content/uploads/2014/02/evohome-controller-display-home-en.png

zxdavb replied:
Regarding the above graphic, I should have said that the quick actions are the operating modes and the overriding decayed is that the state.

balloob replied:
We will not accept any PR that does not use the states provided by the climate component.

Because not using the built-in states means it doesn't work with Alexa, Google Assistant or any other system that will integrate with us, like Web of Things in the future.

zxdavb replied:
I am now using HA states.

jeradM commented on homeassistant/components/evohome.py:
```diff
@@ -0,0 +1,667 @@
+CONF_LOCATION_IDX = 'location_idx'
+
+# Validation of the user's configuration.
+CV_FLOAT = vol.All(vol.Coerce(float), vol.Range(min=MIN_TEMP, max=MAX_TEMP))
```
What is this doing?

zxdavb replied:
Here, nothing - so I have taken it out.  However, it will be going back in later...

This PR was originally 2,500 lines. Balloob asked me to reduce it to the bare minimum for the initial PR, then add back some of the functionality in future PRs.  

FWIW, one example of using this: the user has the opportunity to tell the component if a non-default away temp is configured, because there is no way of extracting that data from the client api.

This is the reason why you might see oddities like the if-if that you rightly suggested could be a single if.

jeradM replied:
:+1: Try to make sure any code that is not used for the items in _this_ PR is all removed. We don't want unused stuff sitting around

zxdavb replied:
Had another tidy-up.

jeradM commented on homeassistant/components/evohome.py:
```diff
@@ -0,0 +1,667 @@
+            'signal': 'update',
+            'to': EVO_MASTER
+        }
+        hass.helpers.dispatcher.async_dispatcher_send(
```
`hass.helpers.dispatcher.dispatcher_send`

zxdavb replied:
Why not async?  I thought async was 'a good thing'.  

I also use `async_schedule_update_ha_state` and `async_dispatcher_connect` - do I switch them over too?

jeradM replied:
Async is a good thing, but you have to make sure you are in the right context for it. Methods and functions that are prefixed `async_` _must_ be run in the event loop. `_first_update` will be run in an executor thread so you need to use the synchronous helpers. Or decorate it `@callback` which will indicate to hass that it is safe to run in the event loop, then `async_` methods are okay. It's all about how things get scheduled. If you haven't seen it already, this might help: https://developers.home-assistant.io/docs/en/asyncio_categorizing_functions.html

zxdavb replied:
Too complicated for me - I will remove the three asyncs for now.

zxdavb replied:
Actually, two - my `_connect()` method has a `@callback`, OK? 
```python
    @callback
    def _connect(self, packet):
        """Process a dispatcher connect."""
        if packet['to'] & self._type and packet['signal'] == 'update':
            self.async_schedule_update_ha_state(force_refresh=True)
```

zxdavb replied:
seems OK to leave this in

jeradM commented on homeassistant/components/evohome.py:
```diff
@@ -0,0 +1,667 @@
+#           pass
+
+# evohomeclient2 (>=0.2.7) now exposes requests exceptions, e.g.:
+# - "400 Client Error: Bad Request for url" (e.g. Bad credentials)
```
No need to describe standard HTTP error codes

zxdavb replied:
What I meant to do here was to document what these HTTP error codes meant in the _context of this application_.

I will change rewrite these comments, putting the emphasis on the text inside the brackets

zxdavb replied:
What I meant to do here was to document what these HTTP error codes meant in the _context of this application_.

I will change rewrite these comments, putting the emphasis on the text inside the brackets

zxdavb replied:
This section has been completely re-written, to address the above, and remove 0-indent comments.

jeradM commented on homeassistant/components/evohome.py:
```diff
@@ -0,0 +1,667 @@
+    def _connect(self, packet):
+        """Process a dispatcher connect."""
+        if packet['to'] & self._type:
+            if packet['signal'] == 'update':
```
These can be a single `if` statement

zxdavb replied:
Done.

jeradM commented on homeassistant/components/evohome.py:
```diff
@@ -0,0 +1,667 @@
+        if not self._available and \
+                self._timers['statusUpdated'] != datetime.min:
+            # this isn't the first (un)available (i.e. after STARTUP), so...
+            _LOGGER.warning(
```
No logging in properties

zxdavb replied:
Can I ask why not? This is a warning, which I feel should be logged. I know that he can tell you whether it devices available or not, but only this morning will tell you _why_ it was unavailable.

Is is it because it is a warning, rather than a debug. Or is there a general rule about doing any logging in properties, as opposed to methods.

jeradM replied:
It's a rule here. I'm still new-ish to the project so I can't give you the ins and outs of it, but really properties should just be returning a value. There shouldn't be any reason to log anything in a property since the value they will return should be determined during the update and logging can be done there if necessary

zxdavb replied:
@jeradM Currently, I am not willing to remove this `_LOGGER.warning()` - how should we resolve the issue?

zxdavb replied:
It now looks like this:
```python
    @property
    def available(self):
        """Return True is the device is available.

        All evohome entities are initially unavailable. Once HA has started,
        state data is then retrieved by the Controller, and then the slaves
        will get a state (e.g. operating_mode, current_temperature).

        However, evohome entities can become unavailable for other reasons.
        """
        return self._available
```

zxdavb replied:
OK, now logs a message only if was previously `self._available = True`.

jeradM commented on homeassistant/components/evohome.py:
```diff
@@ -0,0 +1,667 @@
+    def supported_features(self):
+        """Get the list of supported features of the Controller."""
+        feats = self._supported_features
+        _LOGGER.debug("supported_features(%s) = %s", self._id, feats)
```
No logging in properties

zxdavb replied:
Removed.

jeradM commented on homeassistant/components/evohome.py:
```diff
@@ -0,0 +1,667 @@
+    def current_operation(self):
+        """Return the operation mode of the evohome entity."""
+        curr_op = self._status['systemModeStatus']['mode']
+        _LOGGER.debug("current_operation(%s) = %s", self._id, curr_op)
```
No logging in properties

zxdavb replied:
Removed

zxdavb committed:
Some code improvements as suggested by JeardM

zxdavb committed:
Merge branch 'dev' of https://github.com/home-assistant/home-assistant into evohome-component

zxdavb committed:
Merge branch 'dev' of https://github.com/home-assistant/home-assistant into evohome-component

zxdavb committed:
Rewrite of exception handler

zxdavb committed:
Removed another unwanted logging in properties

zxdavb committed:
Remove async_ version of various methods

zxdavb committed:
moved available heuristics to update()

zxdavb committed:
Merge branch 'dev' of https://github.com/home-assistant/home-assistant into evohome-component

zxdavb commented:
@jeradM once Travis is finished, perhaps you could have another look? ☺️

zxdavb committed:
Cleanup of code, and re-work linter hints

zxdavb committed:
Merge branch 'dev' of https://github.com/home-assistant/home-assistant into evohome-component

balloob commented on homeassistant/components/climate/evohome.py:
```diff
@@ -0,0 +1,47 @@
+    client = hass.data[DATA_EVOHOME]['client']
+    loc_idx = hass.data[DATA_EVOHOME]['params'][CONF_LOCATION_IDX]
+
+# Collect the (master) controller - evohomeclient has no defined way of
```
Although not necessary for this PR, you should try to make this data available in evohomeclient without requiring protected access.

zxdavb replied:
The code owner has previously been reluctant to go for multiple locations.

I'll have a chat with him about it, he's accepted few PRs off me so by presenting him with another PR he may well go for it.

balloob commented on homeassistant/components/evohome.py:
```diff
@@ -0,0 +1,613 @@
+
+These config parameters are presented with their default values:
+
+# scan_interval: 180     # seconds, you can probably get away with 60
```
We shouldn't expose scan interval. We should pick one that makes sense and stick with it.

zxdavb replied:
I think that the existing Honeywell component has a configurable scanning table?

Anyway I'm happy to do this. Accepted wisdom might say 300 seconds but would you be OK if I did 180 seconds?

This would be well below the API rate limit.

zxdavb replied:
OK, set to 180 seconds.  I left the backoff in, as api rate can be exceeded for other reasons.

zxdavb committed:
fixed a minor documentation typo

zxdavb committed:
Merge branch 'dev' of https://github.com/home-assistant/home-assistant into evohome-component

zxdavb committed:
scan_interval is now no longer a configurable option

balloob commented on homeassistant/components/evohome.py:
```diff
@@ -0,0 +1,606 @@
+# Usually, only the controller does client api I/O during update() to pull
+# current state - the exception is when zones pull their own schedules.
+# However, any entity can call methods that will (eventually) change state.
+PARALLEL_UPDATES = 0
```
This only works for platforms.

zxdavb replied:
I can take it out but I will have to put it back in in an upcoming PR.

When fully implemented, the Controller will pull state during its `update()`, and the other entities will use that data during their `update()` without having to use the client library.

That's why the comment what is does...

zxdavb replied:
Moved to `./climate/evohome.py`.

balloob commented on homeassistant/components/evohome.py:
```diff
@@ -0,0 +1,606 @@
+
+# bit masks for dispatcher packets
+EVO_MASTER = 0x01
+EVO_SLAVE = 0x02
```
We should not use master/slave. Use other words like coordinator

zxdavb replied:
OK, switched to `PARENT`, `CHILD` (mainly as same length text).

balloob commented on homeassistant/components/evohome.py:
```diff
@@ -0,0 +1,606 @@
+        logging.getLogger().setLevel(log_level)
+
+    except requests.RequestException as err:
+        if str(HTTP_BAD_REQUEST) in str(err):
```
this feels weird, better to check `if err.response and err.response.status_code == 400:`

zxdavb replied:
Ha! The main reason for this is that I didn't know any Python before 3mo ago (and barely any programming).  I will fix it.

zxdavb replied:
sorted

balloob commented on homeassistant/components/evohome.py:
```diff
@@ -0,0 +1,606 @@
+
+    # IndexError: configured location index is outside the range
+    except IndexError:
+        _LOGGER.warning(
```
We should not recover and do something else. Instead we should fail setup. This is clearly a misconfiguration. 

Imagine removing a site from your account and all of a sudden all entities are replaced with something else. As a user you will expect it to fail.

zxdavb replied:
Fair enough.

zxdavb replied:
OK to do this?
```python
    except IndexError:
        _LOGGER.warning(
            "setup(): Config parameter, '%s'= %s , is out of range (0-%s)",
            CONF_LOCATION_IDX,
            loc_idx,
            len(client.installation_info) - 1
        )

        raise
```

zxdavb replied:
OK, leaving the `warning()` in, as the Traceback is pretty indecipherable without it.

zxdavb replied:
in the end, this was taken out

balloob commented on homeassistant/components/evohome.py:
```diff
@@ -0,0 +1,606 @@
+    return True
+
+
+class EvoEntity(ClimateDevice):
```
Since this one is only supposed to be used inside the evohome climate platform, we should move it to the climate platform.

zxdavb replied:
OK, but I am going to submit subsequent PRs that will split them out again!

zxdavb replied:
was moved.

zxdavb committed:
Change from Master/Slave to Parent/Child

balloob commented on homeassistant/components/evohome.py:
```diff
@@ -0,0 +1,606 @@
+
+        # Set the entity's name & icon (treated as static vales)
+        self._name = domain_data['config']['locationInfo']['name']
+        self._icon = "mdi:thermostat"
```
There is no need for this, all climate devices have this icon. Generally, entities should not write their own icons.

zxdavb replied:
The eventual component will have controller, thermostats, and DHW - I intend to use a different icon for each.

balloob commented on homeassistant/components/evohome.py:
```diff
@@ -0,0 +1,606 @@
+
+        # Set the entity's (initial) behaviour
+        self._available = False  # will be True after first update()
+        self._should_poll = True
```
This should be static as polling is only turned on/off when an entity is added. It might be picked up later if it starts returning True but there is no guarantee.

zxdavb replied:
Other entities will be `self._should_poll = False`, you just wont see them until a later PR.

zxdavb replied:
removed

balloob commented on homeassistant/components/evohome.py:
```diff
@@ -0,0 +1,606 @@
+    @callback
+    def _connect(self, packet):
+        """Process a dispatcher connect."""
+        if packet['to'] & self._type and packet['signal'] == 'update':          # noqa: E501; pylint: disable=no-member
```
What is not a member here and why not?

balloob commented on homeassistant/components/evohome.py:
```diff
@@ -0,0 +1,606 @@
+    def available(self):
+        """Return True is the device is available.
+
+        All evohome entities are initially unavailable. Once HA has started,
```
Why wait till HA starts?

zxdavb replied:
Reworked.

zxdavb commented on homeassistant/components/evohome.py:
```diff
@@ -0,0 +1,606 @@
+        calling AutoWithReset, the controller will enter Auto mode.
+        """
+        if self._status['systemModeStatus']['mode'] == EVO_RESET:
+            state = EVO_AUTO
```
Is your final position that these states can never be, or can only be after home-assistant/architecture/issues/22 is resolved?

TBH, I never thought the controller should be a climate entity anyway, more like a scene...

zxdavb replied:
Complied.

zxdavb committed:
Removed the last of the slaves

zxdavb committed:
Removed the last of the masters

zxdavb committed:
Move -PARALLEL_UPDATES to .\climate\evohome.py'

zxdavb committed:
main code moved to climate/evohome.py

zxdavb committed:
merge EvoEntity into EvoController class

zxdavb committed:
remove should_poll (for now)

zxdavb committed:
woops! left a hint in

zxdavb committed:
removed icon

zxdavb commented on homeassistant/components/climate/evohome.py:
```diff
@@ -0,0 +1,74 @@
+_LOGGER = logging.getLogger(__name__)
+
+#TBD: change LOGGER.info to LOGGER.debug
+def setup_platform(hass, config, add_entities, discovery_info=None):
```
fixed

zxdavb commented on homeassistant/components/climate/evohome.py:
```diff
@@ -0,0 +1,74 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+#TBD: change LOGGER.info to LOGGER.debug
```
fixed

zxdavb commented on homeassistant/components/evohome.py:
```diff
@@ -0,0 +1,1941 @@
+        temperature, which would be a function of operating mode (both
+        controller and zone) and, for TRVs, the OpenWindowMode feature.
+
+        Boilers do not have setpoints; they are only on or off.  Their 
```
fixed

zxdavb committed:
only log a WARNING the first time available = False

zxdavb committed:
cleanup dodgy exception handling

MartinHjelmare commented on homeassistant/components/evohome.py:
```diff
@@ -0,0 +1,193 @@
+            len(client.installation_info) - 1
+        )
+
+        raise
```
Don't raise, `return False`.

zxdavb replied:
Good idea Done.

MartinHjelmare commented on homeassistant/components/evohome.py:
```diff
@@ -0,0 +1,193 @@
+        tmp = dict(domain_data['config'])
+        tmp['locationInfo']['postcode'] = 'REDACTED'
+
+        _LOGGER.debug("setup(): domain_data['config']: %s", tmp)
```
Seems excessive.

zxdavb replied:
De-excessified.

MartinHjelmare commented on homeassistant/components/evohome.py:
```diff
@@ -0,0 +1,193 @@
+
+        _LOGGER.debug("setup(): domain_data['config']: %s", tmp)
+
+    # We have the platofrom configuration, but no state as yet, so...
```
platform

zxdavb replied:
Done

MartinHjelmare commented on homeassistant/components/evohome.py:
```diff
@@ -0,0 +1,193 @@
+    # We have the platofrom configuration, but no state as yet, so...
+    def _first_update(event):
+        """Let the controller know it can obtain it's first update."""
+    # Send a message to the hub to do its first update()
```
Remove this.

zxdavb replied:
Tidied up.  Removed triple-quotes as private method.

MartinHjelmare reviewed (changes requested):
Please minimize logging and comments.

zxdavb committed:
Tidy up exception handling

zxdavb commented on homeassistant/components/climate/evohome.py:
```diff
@@ -0,0 +1,463 @@
+zones (e.g. TRVs, relays) and, optionally, a DHW controller.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/evohome/
```
Changed to `https://home-assistant.io/components/climate.evohome/`

zxdavb commented on homeassistant/components/evohome.py:
```diff
@@ -0,0 +1,193 @@
+                "Unable to continue. Resolve any errors and restart HA.",
+                domain_data['params'][CONF_USERNAME]
+            )
+        else:
```
Yep. Done.

zxdavb commented on homeassistant/components/climate/evohome.py:
```diff
@@ -0,0 +1,463 @@
+    SUPPORT_OPERATION_MODE,
+    SUPPORT_AWAY_MODE,
+)
+
```
Done.

zxdavb commented on homeassistant/components/climate/evohome.py:
```diff
@@ -0,0 +1,463 @@
+    master = EvoController(hass, client, tcs_obj_ref)
+    add_entities([master], update_before_add=False)
+
+    return True
```
removed

zxdavb commented on homeassistant/components/climate/evohome.py:
```diff
@@ -0,0 +1,463 @@
+        Most read-only properties are set here.  So are pseudo read-only,
+        for example name (which _could_ change between update()s).
+        """
+        # Set the usual object references
```
removed this one, and will trawl through the rest.

zxdavb commented on homeassistant/components/climate/evohome.py:
```diff
@@ -0,0 +1,463 @@
+        domain_data = hass.data[DATA_EVOHOME]
+
+        self._obj = obj_ref
+        self._type = EVO_PARENT
```
Can I leave it in?  It is not a constant.  There will other entity types in future PRs..

MartinHjelmare replied:
Then we can change it at that point. We should never add code that we don't need right now.

zxdavb replied:
Removed. You guys are ruthless.

zxdavb commented on homeassistant/components/climate/evohome.py:
```diff
@@ -0,0 +1,463 @@
+
+    @property
+    def available(self):
+        """Return True is the device is available.
```
fixed

zxdavb commented on homeassistant/components/climate/evohome.py:
```diff
@@ -0,0 +1,463 @@
+                 for zone in self._status['zones']]
+
+        avg_temp = sum(temps) / len(temps) if temps else None
+        return round(avg_temp, 1)
```
Fixed

zxdavb commented on homeassistant/components/climate/evohome.py:
```diff
@@ -0,0 +1,463 @@
+        temps = [zone['temperatureStatus']['temperature'] for zone in tmp_list]
+
+        avg_temp = sum(temps) / len(temps) if temps else None
+        return round(avg_temp, 1)
```
Also fixed.

zxdavb commented on homeassistant/components/climate/evohome.py:
```diff
@@ -0,0 +1,463 @@
+    def set_operation_mode(self, operation_mode):
+        """Set new target operation mode for the TCS.
+
+        'AutoWithReset may not be a mode in itself: instead, it _should_(?)
```
Apostrophe found.

zxdavb commented on homeassistant/components/climate/evohome.py:
```diff
@@ -0,0 +1,463 @@
+        return MAX_TEMP
+
+    @property
+    def state(self):
```
`state` property removed & `in_on` added.

zxdavb committed:
Many changes as suggested by @MartinHjelmare, thanks

zxdavb commented on homeassistant/components/climate/evohome.py:
```diff
@@ -0,0 +1,463 @@
+        self._obj = obj_ref
+        self._type = EVO_PARENT
+
+        # Set the entity's own object reference & identifier
```
Gone, generally tidied up too.

zxdavb commented on homeassistant/components/climate/evohome.py:
```diff
@@ -0,0 +1,463 @@
+
+        # Set the entity's operation list - hard-coded for a particular order,
+        # instead of using self._config['allowedSystemModes']
+        self._op_list = TCS_MODES
```
Done.

zxdavb replied:
I think this will be reworked went I sort out HA standard state - anyway removed )op_list and moved constant to `def operation_list()`

zxdavb commented on homeassistant/components/climate/evohome.py:
```diff
@@ -0,0 +1,463 @@
+
+    def turn_away_mode_on(self):
+        """Turn away mode on."""
+        self.set_operation_mode(EVO_AWAY)
```
@MartinHjelmare Sorry, I am a self-taught Python programmer with 3 month's experience - what do you mean?  I see this code in `climate\__init__.py` - isn't that what you mean by base mode?
```python
    def turn_away_mode_on(self):
        """Turn away mode on."""
        raise NotImplementedError()
...
    def turn_away_mode_off(self):
        """Turn away mode off."""
        raise NotImplementedError()
```

MartinHjelmare replied:
`EVO_AWAY` isn't a mode defined in the base climate component. We can't use the `set_operation` method to set any other modes than modes defined in the base climate component. Import the modes needed from the climate component.

zxdavb replied:
I must be thick - I can't see STATE_AWAY in the climate component!

MartinHjelmare replied:
We don't need to use `set_operation` to set away mode. Call the evo api directly.

zxdavb replied:
I was think away mode was another operating_mode.

zxdavb commented on homeassistant/components/climate/evohome.py:
```diff
@@ -0,0 +1,463 @@
+    loc_idx = hass.data[DATA_EVOHOME]['params'][CONF_LOCATION_IDX]
+
+# Collect the (master) controller - evohomeclient has no defined way of
+# accessing non-default location other than using the protected member
```
This one is fixed.

zxdavb committed:
remove hass from init, part 1

MartinHjelmare commented on homeassistant/components/climate/evohome.py:
```diff
@@ -0,0 +1,463 @@
+            )
+
+            try:
+                self._obj._set_status(operation_mode)                           # noqa: E501; pylint: disable=protected-access
```
See eg:
https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/climate/tuya.py#L22-L35

zxdavb replied:
OK, got it.  Is there no STATE_AWAY?  In `climate\__init__.py`, I see:
```
STATE_HEAT = 'heat'
STATE_COOL = 'cool'
STATE_IDLE = 'idle'
STATE_AUTO = 'auto'
STATE_MANUAL = 'manual'
STATE_DRY = 'dry'
STATE_FAN_ONLY = 'fan_only'
STATE_ECO = 'eco'
STATE_ELECTRIC = 'electric'
STATE_PERFORMANCE = 'performance'
STATE_HIGH_DEMAND = 'high_demand'
STATE_HEAT_PUMP = 'heat_pump'
STATE_GAS = 'gas'
```
and I guess I can  also use:
```
STATE_ON, STATE_OFF, STATE_UNKNOWN
```

MartinHjelmare replied:
I don't think there's a `STATE_AWAY`. Yes, we can use the general on/off states too. Don't use `STATE_UNKNOWN`. Use `None` to represent unknown state.

zxdavb replied:
OK, that's what confused me - so for Away mode, I'll use current_op = Auto, and set is_away_mod_one = True.

zxdavb replied:
This is what I got: 
```
HA_STATE_TO_EVO = {
    EVO_RESET = STATE_AUTO,
    EVO_AUTO = STATE_AUTO,
    EVO_AUTOECO = STATE_EVO,
    EVO_AWAY = STATE_AUTO,
    EVO_DAYOFF = STATE_AUTO,
    EVO_CUSTOM = STATE_AUTO,
    EVO_HEATOFF = STATE_OFF
}

HA_STATE_TO_EVO = {
    STATE_AUTO = EVO_AUTO,
    STATE_EVO = EVO_AUTOECO,
    STATE_OFF = EVO_HEATOFF
}
```

MartinHjelmare replied:
I guess it should be `STATE_ECO` and not `STATE_EVO` in the second dictionary.

zxdavb committed:
use async_added_to_hass instead of dispatcher_connect

zxdavb committed:
remove hass from init, part 2 (done)

zxdavb commented on homeassistant/components/climate/evohome.py:
```diff
@@ -0,0 +1,463 @@
+    the child (CH/DHW) devices.
+    """
+
+    def __init__(self, hass, client, obj_ref):
```
`hass` is gone.

zxdavb committed:
add 1-2 arrays, and tidied up some comments

zxdavb committed:
from dispatcher to async_added_to_hass

zxdavb commented on homeassistant/components/climate/evohome.py:
```diff
@@ -0,0 +1,463 @@
+        if not expired:  # timer not expired, so exit
+            return True
+
+        # If we reached here, then it is time to update state data.  NB: unlike
```
Removed the comment.

zxdavb commented on homeassistant/components/climate/evohome.py:
```diff
@@ -0,0 +1,463 @@
+            # only warn if available from False to True
+            _LOGGER.warning(
+                "The entity, %s, is now unavailable "
+                "(i.e. self.available() = False), debug code is: %s",
```
Resolved.

zxdavb commented on homeassistant/components/climate/evohome.py:
```diff
@@ -0,0 +1,463 @@
+            # this isn't the first re-available (e.g. _after_ STARTUP)
+            _LOGGER.debug(
+                "The entity, %s, is now available "
+                "(i.e. self.available() = True)",
```
the comment is explaining the `elif` clause (i.e. the significant of `datetime.min`), the debug message doesn't include the _after_ STARTUP qualifier - I would like to leave this in.

zxdavb commented on homeassistant/components/evohome.py:
```diff
@@ -0,0 +1,193 @@
+        tmp[CONF_USERNAME] = 'REDACTED'
+        tmp[CONF_PASSWORD] = 'REDACTED'
+
+        _LOGGER.debug("setup(): Configuration parameters: %s", tmp)
```
Fair enough - removed.

zxdavb commented on homeassistant/components/climate/evohome.py:
```diff
@@ -0,0 +1,463 @@
+                self._id + " [" + self._name + "]"
+            )
+
+        return True
```
done

zxdavb committed:
cleaned up some logging, and removed others

zxdavb commented on homeassistant/components/climate/evohome.py:
```diff
@@ -0,0 +1,463 @@
+        )
+
+        try:
+            domain_data['status'].update(  # or: domain_data['status'] =
```
removed. it was a note to myself

zxdavb commented on homeassistant/components/climate/evohome.py:
```diff
@@ -0,0 +1,463 @@
+            self._connect
+        )  # for: def async_dispatcher_connect(signal, target)
+
+        if _LOGGER.isEnabledFor(logging.DEBUG):
```
I have left just 1 DEBUG line, I would really prefer to keep that one in.

zxdavb commented on homeassistant/components/climate/evohome.py:
```diff
@@ -0,0 +1,463 @@
+        self._available = False  # will be True after first update()
+
+        # Create a listener for (internal) update packets...
+        hass.helpers.dispatcher.dispatcher_connect(
```
Done - I have seen your comments, below - I think I misunderstood.

zxdavb commented on homeassistant/components/climate/evohome.py:
```diff
@@ -0,0 +1,463 @@
+EVO_HEATOFF = 'HeatingOff'
+
+TCS_MODES = [
+    EVO_RESET,
```
I am a little confused about STATEs and MODEs.  Is STATE_AWAY an official state?

zxdavb replied:
Done

zxdavb commented on homeassistant/components/climate/evohome.py:
```diff
@@ -0,0 +1,463 @@
+            timedelta(seconds=domain_data['params'][CONF_SCAN_INTERVAL])
+
+        if not expired:  # timer not expired, so exit
+            return True
```
removed True

zxdavb commented on homeassistant/components/climate/evohome.py:
```diff
@@ -0,0 +1,463 @@
+        if no_recent_updates:
+            # unavailable because no successful update()s (but why?)
+            self._available = False
+            debug_code = '0x01'
```
Done

zxdavb commented on homeassistant/components/evohome.py:
```diff
@@ -0,0 +1,193 @@
+Support for a temperature control system (TCS, controller) with 0+ heating
+zones (e.g. TRVs, relays) and, optionally, a DHW controller.
+
+A minimal configuration.yaml is as as below:
```
Removed - I will checked the docs PR & updated it - will push it just now

zxdavb commented on homeassistant/components/evohome.py:
```diff
@@ -0,0 +1,193 @@
+    # Used for internal data, such as installation, state & timers...
+    hass.data[DATA_EVOHOME] = {}
+
+    domain_data = hass.data[DATA_EVOHOME]
```
Beautiful.

zxdavb commented on homeassistant/components/evohome.py:
```diff
@@ -0,0 +1,193 @@
+
+    try:
+        # there's a bug in evohomeclient2 v0.2.7: the client.__init__() sets
+        # the root loglevel (debug=?), so must remember it now...
```
I have had it fixed (it had made my life hell) - it is not released to PyPi as yet (there is no 0.2.8 as yet).  I have other bugs for the client library, but I just want this PR accepted first.

zxdavb commented on homeassistant/components/evohome.py:
```diff
@@ -0,0 +1,193 @@
+            )
+        else:
+            # Otherwise, it may be enough to back off and try again later.
+            raise PlatformNotReady(err)
```
TBH, I am not sure about the difference between a platform and whatever.  I put it in on someone's suggestion - can take it out?

zxdavb replied:
@MartinHjelmare thanks for your help so far - can you explain this one to me?  

The vendor's websites are unreliable, and I often got 503s when starting HA - the idea was to retry a few times, hoping things would resolve (they invariably do, given time).

zxdavb replied:
Currently, the code looks like:
```python
    except requests.exceptions.HTTPError as err:
        if err.response.status_code == HTTP_BAD_REQUEST:
            _LOGGER.error(
                "Failed to establish a connection with evohome web servers, "
                "Check your username (%s), and password are correct."
                "Unable to continue. Resolve any errors and restart HA.",
                domain_data['params'][CONF_USERNAME]
            )
            return False  # unable to continue

        raise PlatformNotReady(err)

    except requests.RequestException as err:
        raise PlatformNotReady(err)
```

zxdavb replied:
conversation continuing below...

zxdavb commented on homeassistant/components/climate/evohome.py:
```diff
@@ -0,0 +1,425 @@
+    TEMP_CELSIUS,
+    HTTP_TOO_MANY_REQUESTS,
+)
+from homeassistant.core import callback
```
sorted

zxdavb commented on homeassistant/components/climate/evohome.py:
```diff
@@ -0,0 +1,425 @@
+    SUPPORT_OPERATION_MODE,
+    SUPPORT_AWAY_MODE,
+)
+from homeassistant.components.evohome import (
```
sorted

zxdavb commented on homeassistant/components/climate/evohome.py:
```diff
@@ -0,0 +1,424 @@
+HA_STATE_TO_EVO = {
+    STATE_AUTO: 'Auto',
+    STATE_ECO: 'AutoWithEco',
+    STATE_OFF: 'HeatingOff'
```
fixed

zxdavb commented on homeassistant/components/climate/evohome.py:
```diff
@@ -0,0 +1,424 @@
+
+HA_STATE_TO_EVO = {
+    STATE_AUTO: 'Auto',
+    STATE_ECO: 'AutoWithEco',
```
fixed

zxdavb commented on homeassistant/components/climate/evohome.py:
```diff
@@ -0,0 +1,424 @@
+]
+
+HA_STATE_TO_EVO = {
+    STATE_AUTO: 'Auto',
```
fixed

zxdavb committed:
Many changes as request by @MartinHjelmare

zxdavb committed:
Homage to the lint

MartinHjelmare commented on homeassistant/components/climate/evohome.py:
```diff
@@ -0,0 +1,425 @@
+    @asyncio.coroutine
+    def async_added_to_hass(self):
+        """Start pulling down state as entity about to be added."""
+        self.async_schedule_update_ha_state(force_refresh=True)
```
Ok. You can decide if you want to delay the update until home assistant start and use the dispatcher to do that.

If we don't need to delay update until home assistant start we don't need the dispatcher or use `async_added_to_hass`. We can just let update be called during entity addition by setting `update_before_add` to `True` in `add_entities`.

zxdavb replied:
I think I'll `update_before_add=True` now - it is the simplest way, and go back to dispatcher/`async_added_to_hass` when zones are added back in.

zxdavb replied:
done

zxdavb committed:
Changed to the HA of doing operating_mode

zxdavb commented on homeassistant/components/climate/evohome.py:
```diff
@@ -0,0 +1,463 @@
+
+    def turn_away_mode_off(self):
+        """Turn away mode off."""
+        self.set_operation_mode(EVO_AUTO)
```
Now doing this:
```python
    def turn_away_mode_off(self):
        """Turn away mode off."""
        self._set_operation_mode(EVO_AUTO)

    def _set_operation_mode(self, operation_mode):
        # Set new target operation mode for the TCS.
        _LOGGER.debug(
            "_set_operation_mode(): API call [1 request(s)]: "
            "tcs._set_status(%s)...",
            operation_mode
        )
        try:
            self._obj._set_status(operation_mode)                                      # noqa: E501; pylint: disable=protected-access
        except HTTPError as err:
            self._handle_requests_exceptions("HTTPError", err)

    def set_operation_mode(self, operation_mode):
        self._set_operation_mode(HA_STATE_TO_EVO.get(operation_mode))
```

zxdavb committed:
Now using update_before_add=True

MartinHjelmare commented on homeassistant/components/climate/evohome.py:
```diff
@@ -0,0 +1,395 @@
+
+from datetime import datetime, timedelta
+import logging
+from requests.exceptions import HTTPError
```
Add a blank line between standard library and 3rd party imports.

zxdavb replied:
done

MartinHjelmare commented on homeassistant/components/climate/evohome.py:
```diff
@@ -0,0 +1,395 @@
+
+        self._available = False  # should become True after first update()
+
+        if _LOGGER.isEnabledFor(logging.DEBUG):
```
Remove this logging.

zxdavb replied:
I want to log the configuration somewhere.  This is done once only, what is wrong with that?

Where else can I put it?

MartinHjelmare replied:
If it's really necessary move it to setup_platform where there already is a debug log.

zxdavb replied:
Ok, I took it out if `__init__()`, simplified logging in `setup_platform()`, and expanded logging in `setup()` a little.

MartinHjelmare commented on homeassistant/components/evohome.py:
```diff
@@ -0,0 +1,146 @@
+            )
+            return False  # unable to continue
+
+        raise PlatformNotReady(err)
```
Remove this. We can't use this in the component.

MartinHjelmare replied:
Should we return False in this case too?

zxdavb replied:
I'm inclined to simply raise.  If I do `return False`, then I should add `_LOGGING.error()`.  Which do you think is better...

MartinHjelmare replied:
We don't raise. Log error and return False.

zxdavb replied:
I only want to handle `HTTP_BAD_REQUEST`, because it is a mis-configuration by the user, and the TraceBack can be a little cryptic (i.e. it is not HTTP_UNAUTHORIZED, or 'bad credentials'). 
 That's why there is an `except` there.  

If I didn't have that `except`, then an exception would occur via the client library.  I am not raising, merely allowing exceptions that I don't want to handle do their natural thing.

I would like: 
```python
    except requests.exceptions.HTTPError as err:
        if err.response.status_code == HTTP_BAD_REQUEST:
            _LOGGER.error(
                "Failed to establish a connection with evohome web servers, "
                "Check your username (%s), and password are correct."
                "Unable to continue. Resolve any errors and restart HA.",
                domain_data['params'][CONF_USERNAME]
            )
            return False  # unable to continue

        raise  # we dont handle any other HTTPErrors
```

zxdavb replied:
FYI, here is what `homeassistant.log` looks like when the above scenario occurs, a `503` in this case, and with my preferred solution:
```
2018-09-25 09:58:16 DEBUG (SyncWorker_12) [homeassistant.components.evohome] setup(): API call [4 request(s)]: client.__init__()...
   ...
2018-09-25 09:58:16 ERROR (MainThread) [homeassistant.setup] Error during setup of component evohome
Traceback (most recent call last):
  File "/home/dbonnes/home-assistant/homeassistant/setup.py", line 148, in _async_setup_component
    component.setup, hass, processed_config)  # type: ignore
  File "/usr/lib/python3.6/concurrent/futures/thread.py", line 56, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/home/dbonnes/home-assistant/homeassistant/components/evohome.py", line 82, in setup
    debug=False
  File "/home/dbonnes/home-assistant/lib/python3.6/site-packages/evohomeclient2/__init__.py", line 16, in __init__
    self._login()
  File "/home/dbonnes/home-assistant/lib/python3.6/site-packages/evohomeclient2/__init__.py", line 82, in _login
    self._basic_login()
  File "/home/dbonnes/home-assistant/lib/python3.6/site-packages/evohomeclient2/__init__.py", line 70, in _basic_login
    r.raise_for_status()
  File "/home/dbonnes/home-assistant/lib/python3.6/site-packages/requests/models.py", line 939, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 503 Server Error: Service Unavailable for url: https://tccna.honeywell.com/Auth/OAuth/Token
```

zxdavb replied:
Assume this is OK.

MartinHjelmare commented on homeassistant/components/evohome.py:
```diff
@@ -0,0 +1,146 @@
+        raise PlatformNotReady(err)
+
+    except requests.RequestException as err:
+        raise PlatformNotReady(err)
```
Remove this and return False instead.

zxdavb replied:
I think better to take the whole `except` clause out...

zxdavb replied:
Have taken this clause out.

zxdavb committed:
reduced logging further still...

zxdavb committed:
Merge branch 'dev' of https://github.com/home-assistant/home-assistant into evohome-component

zxdavb committed:
fixed minor lint

zxdavb commented on homeassistant/components/evohome.py:
```diff
@@ -0,0 +1,145 @@
+    HTTP_BAD_REQUEST
+)
+
+from homeassistant.exceptions import PlatformNotReady
```
fixed

zxdavb committed:
fix a small logic error

zxdavb committed:
Add device_state_attributes to track actual operating mode

zxdavb committed:
Clean up doc quotes caused by  previous changes

zxdavb committed:
Woops! removed some debug lines that shoudln't have stayed in

zxdavb committed:
Add a complete set of device_state_attributes

zxdavb committed:
Cleanup some constants

zxdavb committed:
Remove more legacy code

zxdavb commented on homeassistant/components/climate/evohome.py:
```diff
@@ -0,0 +1,375 @@
+    STATE_OFF: EVO_HEATOFF
+}
+
+HA_OP_LIST = list(HA_STATE_TO_EVO.keys())
```
I read that - thought leaving keys in was more readable - what about: `HA_OP_LIST = list(*HA_STATE_TO_EVO)`

MartinHjelmare replied:
```py
HA_OP_LIST = list(HA_STATE_TO_EVO)
```

zxdavb replied:
Is done.

zxdavb committed:
Merge branch 'dev' of https://github.com/home-assistant/home-assistant into evohome-component

zxdavb committed:
domain_data to evo_data & this else should be a finally

zxdavb commented on homeassistant/components/climate/evohome.py:
```diff
@@ -0,0 +1,375 @@
+
+    Here, we add the controller only.
+    """
+    domain_data = hass.data[DATA_EVOHOME]
```
OK, good suggestion - done.

zxdavb committed:
minor change for readability

zxdavb committed:
Minor change for readability #2

zxdavb committed:
removed unused code

zxdavb committed:
small tidy up - no code changes

MartinHjelmare reviewed (approved):
I think this is ok now. There are still some remaining `if _LOGGER.isEnabledFor(logging.DEBUG):` which I don't think is necessary, but nothing major in the way of merging in my opinion.

MartinHjelmare commented on homeassistant/components/evohome.py:
```diff
@@ -0,0 +1,144 @@
+zones (e.g. TRVs, relays) and, optionally, a DHW controller.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/climate.evohome/
```
URL is wrong.

zxdavb replied:
What should it be? `https://www.home-assistant.io/components/climate.evohome/`?

zxdavb replied:
The web page doesn't exist yet - it is a PR - home-assistant.io/6252

MartinHjelmare replied:
It should be `https://home-assistant.io/components/evohome/`. It's ok to add it here before the PR is merged.

MartinHjelmare replied:
Just to be clear, there should be two pages, one for the component and one for the platform.

zxdavb replied:
Ah!

zxdavb replied:
OK, done.

MartinHjelmare commented on homeassistant/components/evohome.py:
```diff
@@ -0,0 +1,144 @@
+#   0-1 DHW controller, (a.k.a. Boiler)
+
+import logging
+from requests.exceptions import HTTPError
```
Add a blank line between standard library and 3rd party imports.

zxdavb replied:
Done

zxdavb committed:
fix minor lint

zxdavb committed:
Correct URLs & descriptions in docstring

MartinHjelmare commented on homeassistant/components/climate/evohome.py:
```diff
@@ -0,0 +1,372 @@
+Support for a temperature control system (TCS, controller) with 0+ heating
+zones (e.g. TRVs, relays) and, optionally, a DHW controller.
+
+For more details about this component, please refer to the documentation at
```
This is the platform.

zxdavb replied:
woops

MartinHjelmare commented on homeassistant/components/evohome.py:
```diff
@@ -0,0 +1,145 @@
+Support for a temperature control system (TCS, controller) with 0+ heating
+zones (e.g. TRVs, relays) and, optionally, a DHW controller.
+
+For more details about this platform, please refer to the documentation at
```
This is the component.

zxdavb replied:
woops - fixed

zxdavb commented:
@MartinHjelmare thanks for all your help

zxdavb committed:
whoops - fixed a typo in docstrings

zxdavb committed:
Merge branch 'dev' of https://github.com/home-assistant/home-assistant into evohome-component

zxdavb commented:
From: https://deploy-preview-92--developers-home-assistant.netlify.com/docs/en/next/integration_quality_scale_index.html

This integration is able to cope when things go wrong. It will not print any exceptions nor will it fill the log with retry attempts.

[x] Set an appropriate SCAN_INTERVAL (if a polling integration) - _180 sec_
[x] Raise PlatformNotReady if unable to connect during platform setup (if appropriate) - N/A (see above)
[x] Handles expiration of auth credentials. Refresh if possible or print correct error and fail setup. If based on a config entry, should trigger a new config entry flow to re-authorize. - _refresh logic is now in client api_
[ ] Handles internet unavailable. Log a warning once when unavailable, log once when reconnected.
[x] Handles device/service unavailable. Log a warning once when unavailable, log once when reconnected - _yes_
[x] Set available property to False if appropriate (docs) - _yes_
[x] Entities have unique ID (if available) - _yes (the api is undocumented, but it is believed the Honeywell-assigned IDs are unique_

MartinHjelmare commented:
Note that the last point means that the `unique_id` property should be defined in the platform entity class.

zxdavb commented:
> Note that the last point means that the `unique_id` property should be defined in the platform entity class.

Ok, I'll take that tick off, and look to achieving Silver in a subsequent PR (should be doable), once I understand exactly you mean, above.

zxdavb committed:
remove an unused line of cide & a small tidy-up

zxdavb committed:
Merge branch 'dev' of https://github.com/home-assistant/home-assistant into evohome-component

balloob commented:
🎉 🎉 🎉 🎉 🎉
